kratos/geometries/nurbs_curve_geometry.h

```diff
@@ -0,0 +1,488 @@
+    {
+        BaseType::operator=(rOther);
+        mPolynomialDegree = rOther.mPolynomialDegree;
+        mKnots = rOther.mKnots;
+        mWeights = rOther.mWeights;
+        return *this;
+    }
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    typename BaseType::Pointer Create(
+        PointsArrayType const& ThisPoints) const override
+    {
+        return Kratos::make_shared<NurbsCurveGeometry>(ThisPoints);
+    }
+
+    ///@}
+    ///@name Get and Set functions
+    ///@{
+
+    /*
+    * @brief Polynomial degree of this curve.
+    * @return the polynomial degree.
+    */
+    SizeType PolynomialDegree() const
+    {
+        return mPolynomialDegree;
+    }
+
+    /* 
+    * @brief Knot vector is defined to have a multiplicity of p
+    *        at the beginning and end (NOT: p + 1).
+    * @return knot vector.
+    */
+    const Vector& Knots() const
+    {
+        return mKnots;
+    }
+
+    /*
+    * @brief The number of knots within the knot vector.
+    * @return the size of the knot vector.
+    */
+    SizeType NumberOfKnots() const
+    {
+        return mKnots.size();
+    }
+
+    /*
+    * @brief The number of nonzero control points for one point in the curve
+    *        is given by p+1.
+    * @return the number of nonzero control points.
+    */
+    SizeType NumberOfNonzeroControlPoints() const
+    {
+        return PolynomialDegree() + 1;
+    }
+
+    /*
+    * @brief Checks if shape functions are rational or not.
+    * @return true if NURBS,
+    *         false if B-Splines only (all weights are considered as 1.0)
+    */
+    bool IsRational() const
+    {
+        return mWeights.size() != 0;
+    }
+
+    /*
+    * @brief Provides weights vector. All values are 1.0 for B-Splines,
+    *        for NURBS those can be unequal 1.0.
+    *        Has size 0 if B-Spline.
+    * @return weights vector.
+    */
+    const Vector& Weights() const
+    {
+        return mWeights;
+    }
+
+    /* 
+    * @brief Provides the natural boundaries of the NURBS/B-Spline curve.
+    * @return domain interval.
+    */
+    Interval DomainInterval() const
+    {
+        return Interval(mKnots[mPolynomialDegree - 1], mKnots[NumberOfKnots() - mPolynomialDegree]);
+    }
+
+    /*
+    * @brief Provides all knot intervals within one curve.
+    * @return vector of domain intervals.
+    */
+    std::vector<Interval> KnotSpanIntervals() const
+    {
+        const IndexType first_span = mPolynomialDegree - 1;
+        const IndexType last_span = NumberOfKnots() - mPolynomialDegree - 1;
+
+        const IndexType number_of_spans = last_span - first_span + 1;
+
+        std::vector<Interval> result(number_of_spans);
+
+        for (IndexType i = 0; i < number_of_spans; i++) {
+            const double t0 = mKnots[first_span + i];
+            const double t1 = mKnots[first_span + i + 1];
+
+            result[i] = Interval(t0, t1);
+        }
+
+        return result;
```

Comment:
would using `std::move` make sense here?

not sure, asking the others

Reply:
copy elision?

Reply:
what do you mean?

Reply:
std::move would make it more explicit (which is good) but as result is rvalue, the optimizer will use the move.